### PR TITLE
Itr fsc

### LINKED ISSUE #19: iterative ref: compute_fsc
* TODO: write fast vectorized fsc from code snippets in 
  *  https://github.com/geoffwoollard/learn_cryoem_math/blob/master/nb/fsc.ipynb
  * https://github.com/geoffwoollard/learn_cryoem_math/blob/master/nb/mFSC.ipynb
  * https://github.com/geoffwoollard/learn_cryoem_math/blob/master/nb/guinier_fsc_sharpen.ipynb

### PR DESCRIPTION
**Pull request by arjunsingh3600**
compute_fsc implementation

### CODE REVIEW

File: tests/test_expectation_maximization.py [hunk 1]
```diff
@@ -250,11 +250,16 @@ def test_insert_slice(test_ir, n_pix):
...
-    map_2 = np.ones((n_pix, n_pix, n_pix))
+    map_1_ones = np.ones((n_pix, n_pix, n_pix))
+
+    fsc_1 = test_ir.compute_fsc(map_1_ones, map_1_ones)
+    fsc_diff_amplitudes = test_ir.compute_fsc(map_1_ones * 2, map_1_ones * 4.5)
+    fsc_diff_phases = test_ir.compute_fsc(map_1_ones * 1, map_1_ones * -1)
 
-    fsc_1 = test_ir.compute_fsc(map_1, map_2)
     assert fsc_1.shape == (n_pix // 2,)
+    assert np.count_nonzero(fsc_1.real != 1) == 0
```
Comment (geoffwoollard) [inline tests/test_expectation_maximization.py]: `np.allclose(fsc_1.real,1)`

File: tests/test_expectation_maximization.py [hunk 2]
```diff
@@ -250,11 +250,16 @@ def test_insert_slice(test_ir, n_pix):
...
+    assert np.count_nonzero(fsc_diff_amplitudes.real != 1) == 0
```
Comment (geoffwoollard) [inline tests/test_expectation_maximization.py]: `np.allclose(fsc_diff_amplitudes.real,1)`

File: tests/test_expectation_maximization.py [hunk 3]
```diff
@@ -250,11 +250,16 @@ def test_insert_slice(test_ir, n_pix):
...
+    assert np.count_nonzero(fsc_diff_phases.real != -1) == 0
```
Comment (geoffwoollard) [inline tests/test_expectation_maximization.py]: np.allclose(fsc_diff_phases.real,-1)

File: tests/test_expectation_maximization.py (hunk 1 above)
Comment (geoffwoollard) [reply tests/test_expectation_maximization.py]: Tests hard to read. Too many double negatives.

File: tests/test_expectation_maximization.py [hunk 4]
```diff
@@ -250,11 +250,16 @@ def test_insert_slice(test_ir, n_pix):
 
 def test_compute_fsc(test_ir, n_pix):
     """Test computation of FSC."""
-    map_1 = np.ones((n_pix, n_pix, n_pix))
-    map_2 = np.ones((n_pix, n_pix, n_pix))
+    map_1_ones = np.ones((n_pix, n_pix, n_pix))
```
Comment (geoffwoollard) [inline tests/test_expectation_maximization.py:351]: these are good tests, since they test expected behaviour.

**Comment by geoffwoollard** [comment]
Should you have `IterativeRefinement.stuff` inside a function (here `compute_fsc`)?

https://github.com/compSPI/reconstructSPI/runs/5880657011?check_suite_focus=true#step:6:79

```
        for rad in range(1, n_pix // 2 + 1):
>           shell_mask = IterativeRefinement.binary_mask_3d(
                center=(n_pix // 2, n_pix // 2, n_pix // 2),
                radius=rad,
                shape=map_3d_f_1.shape,
                fill=False,
            ).astype(np.bool)
E           AttributeError: type object 'IterativeRefinement' has no attribute 'binary_mask_3d'

reconstructSPI/iterative_refinement/expectation_maximization.py:828: AttributeError
```

**Comment by arjunsingh3600** [comment]
Looking into `IterativeRefinement.binary_mask_3d()` test fail rn

**Comment by geoffwoollard** [comment]
It should be OK. It's used this way in other functions
https://github.com/compSPI/reconstructSPI/blob/dev/reconstructSPI/iterative_refinement/expectation_maximization.py#L735
https://github.com/compSPI/reconstructSPI/blob/dev/reconstructSPI/iterative_refinement/expectation_maximization.py#L799
https://github.com/compSPI/reconstructSPI/blob/dev/reconstructSPI/iterative_refinement/expectation_maximization.py#L909
https://github.com/compSPI/reconstructSPI/blob/dev/reconstructSPI/iterative_refinement/expectation_maximization.py#L919

**Comment by geoffwoollard** [comment]
It seems your branch is behind dev. why don't you try to merge in dev, or start a fresh branch off dev and put in your additions (open another PR)?

Maybe it's some issue with` binary_mask_3d` versus `binary_mask`

**Comment by arjunsingh3600** [comment]
@geoffwoollard I'll try that out, thanks!

**Comment by geoffwoollard** [comment]
Also check that dev is not broken!

**Comment by arjunsingh3600** [comment]
@geoffwoollard should be good to merge

**Review by geoffwoollard — APPROVED** [review]

**Comment by geoffwoollard** [comment]
Closes https://github.com/compSPI/reconstructSPI/issues/19

**Pull request merged**